src/inference/src/ie_core.cpp

```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
+        try {
+            // do not reshape/re-batch originally batched networks and when there are no inputs with the N* layouts
+            // the below code is a placeholder for the WIP (22.1) functionality
+            // that will check the reshaping by the batch is robust (CVS-51744)
+            const InputsDataMap inputInfo = network.getInputsInfo();
+            bool atLeastOneInputIsBatched = false;
+            for (const InputsDataMap::value_type& item : inputInfo) {
+                auto layout = item.second->getTensorDesc().getLayout();
+                if (layout == InferenceEngine::Layout::NC || layout == InferenceEngine::Layout::NCDHW ||
+                    layout == InferenceEngine::Layout::NCHW || layout == InferenceEngine::Layout::NHWC ||
+                    layout == InferenceEngine::Layout::NDHWC) {
+                    if (item.first == "im_info"  // WA for faster-rcnn* and alikes (CVS-74085)
+                        || 1 != item.second->getTensorDesc().getDims()[0])  // do not reshape/re-batch  batched networks
+                        IE_THROW(NotImplemented)
+                            << "Auto-batching does not reshape/re-batch originally batched networks!";
+                    else
+                        atLeastOneInputIsBatched = true;
+                }
+            }
+            bool atLeastOneOutputIsBatched = false;
+            const OutputsDataMap outputInfo = network.getOutputsInfo();
+            for (const OutputsDataMap::value_type& item : outputInfo) {
+                auto layout = item.second->getTensorDesc().getLayout();
+                if (layout == InferenceEngine::Layout::NC || layout == InferenceEngine::Layout::NCDHW ||
+                    layout == InferenceEngine::Layout::NCHW || layout == InferenceEngine::Layout::NHWC ||
+                    layout == InferenceEngine::Layout::NDHWC) {
+                    if (1 != item.second->getTensorDesc()
+                                 .getDims()[0])  // do not reshape/re-batch originally batched networks
+                        IE_THROW(NotImplemented)
+                            << "Auto-batching does not reshape/re-batch originally batched networks!";
+                    else
+                        atLeastOneOutputIsBatched = true;
+                }
+            }
+            if (!atLeastOneInputIsBatched || !atLeastOneOutputIsBatched)
+                IE_THROW(NotImplemented)
+                    << "Auto-batching supports only networks featuring inputs/outputs with the batched layouts !";
+
+            unsigned int requests = 0;
+            unsigned int optimalBatchSize = 0;
+            if (deviceNameWithBatchSize.empty()) {
+                // batch size is not set explicitly via device name e.g. BATCH:GPU(4)
+                // let's query the optimal batch size
+                std::map<std::string, ie::Parameter> options;
+                options["MODEL_PTR"] = std::const_pointer_cast<ngraph::Function>(network.getFunction());
```

Comment:
why do we need to `const_pointer_cast`? I suppose plugin can analyze the const function as well

Reply:
this is to mimic the behavior of the MAX_BATCH_SIZE, which had a heated discussion, and this is general limitation of ov::Parameter that it seperates const vs non const ( so you will have a cast in one case anyway)